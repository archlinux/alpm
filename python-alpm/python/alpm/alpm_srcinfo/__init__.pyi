"""A module for parsing and linting of ALPM SRCINFO files."""

from . import error, source_info
from .error import SourceInfoError
from .source_info.v1 import SourceInfoV1
from .source_info.v1.merged import MergedPackage

__all__ = [
    "SourceInfoError",
    "error",
    "source_info",
    "SourceInfoV1",
    "MergedPackage",
]
