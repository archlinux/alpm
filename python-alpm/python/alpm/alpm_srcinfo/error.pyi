"""Exceptions used by alpm_srcinfo module."""

class SourceInfoError(Exception):
    """The high-level exception that can occur when using alpm_srcinfo module."""

__all__ = [
    "SourceInfoError",
]
