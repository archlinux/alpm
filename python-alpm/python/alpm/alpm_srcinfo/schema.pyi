"""Schemas for SRCINFO data."""

from pathlib import Path
from typing import Union, TYPE_CHECKING

if TYPE_CHECKING:
    from alpm.alpm_types import SchemaVersion

class SourceInfoSchema:
    """SRCINFO schema.

    The schema of a SRCINFO refers to the minimum required sections and keywords, as
    well as the complete set of available keywords in a specific version.
    """

    def __init__(self, version: Union[str, "SchemaVersion"]) -> None:
        """Create a SourceInfoSchema from SchemaVersion or its string representation.

        Args:
            version (Union[str, SchemaVersion]): either a SchemaVersion or a string
                representation of SchemaVersion.

        Raises:
            ALPMError: if the string representation of the version is invalid.
            SourceInfoError: if there is no corresponding schema for the provided major
                version.

        """

    @staticmethod
    def derive_from_str(srcinfo: str) -> "SourceInfoSchema":
        """Derive the schema from a string containing SRCINFO data.

        Args:
            srcinfo (str): The srcinfo string to derive the schema from.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the srcinfo string cannot be parsed.

        """

    @staticmethod
    def derive_from_file(path: Union[str, Path]) -> "SourceInfoSchema":
        """Derive the schema from a file containing SRCINFO data.

        Args:
            path (Union[str, Path]): The path to the file containing SRCINFO data.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the file cannot be read or parsed.

        """

    @property
    def version(self) -> "SchemaVersion":
        """The schema version."""

    def __str__(self) -> str: ...
    def __repr__(self) -> str: ...
    def __eq__(self, other: object) -> bool: ...

__all__ = [
    """SourceInfoSchema""",
]
