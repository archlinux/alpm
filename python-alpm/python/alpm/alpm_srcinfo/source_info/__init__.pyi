"""Data representations and integrations for reading of SRCINFO data."""

from . import v1

__all__ = ["v1"]
