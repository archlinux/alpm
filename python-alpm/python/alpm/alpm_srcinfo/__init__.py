# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm import alpm_srcinfo as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
