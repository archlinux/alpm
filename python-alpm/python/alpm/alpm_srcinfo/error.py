# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm.alpm_srcinfo import error as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
