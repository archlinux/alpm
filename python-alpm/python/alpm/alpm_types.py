# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from typing import Union

from typing_extensions import TypeAlias

from .alpm import alpm_types as _mod

Checksum: TypeAlias = Union[
    _mod.Blake2b512Checksum,
    _mod.Md5Checksum,
    _mod.Sha1Checksum,
    _mod.Sha224Checksum,
    _mod.Sha256Checksum,
    _mod.Sha384Checksum,
    _mod.Sha512Checksum,
]

__all__ = [*_mod.__all__, "Checksum"]
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
