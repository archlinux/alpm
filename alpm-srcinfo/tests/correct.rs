use std::{fs::read_to_string, path::PathBuf};

use alpm_srcinfo::{MergedPackage, SourceInfo};
use alpm_types::Architecture;
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// Get some correct SRCINFO files and make sure the JSON output is created as expected.
///
/// This test also looks for specific keywords in the generated output, specifically:
/// - `unexpected` is used for any kind of value that shouldn't be included in the JSON output.
/// - `beefc0ffee` is used to mark hex values that shouldn't be included in the JSON.
///
/// The SRCINFO files are generated from `*.pkgbuild` files in the `tests/correct` folder
/// Each `*.pkgbuild` file contains an explanation of what it tests.
/// To regenerate the SRCINFO files run the following command in the `alpm-srcinfo` project root:
///
/// ```sh
/// ./tests/generate_srcinfo.bash tests/*/*.pkgbuild
/// ```
///
/// `makepkg` expects changelog and INSTALL files to be in the build directory when creating
/// the SRCINFO file. The script also takes care of creating those files.
#[rstest]
pub fn correct_files(#[files("tests/correct/*.srcinfo")] case: PathBuf) -> TestResult {
    // Read the input file and parse it.
    let input = read_to_string(&case)?;
    let source_info_result = SourceInfo::from_string(input.as_str());

    // Make sure there're no parse errors
    let source_info_result = match source_info_result {
        Ok(result) => result,
        Err(err) => {
            return Err(format!(
                "The parser errored even though it should've succeeded the parsing step:\n{err}"
            )
            .into());
        }
    };

    // Ensure that there're lint errors
    let source_info = match source_info_result.lint() {
        Ok(source_info) => source_info,
        Err(err) => {
            return Err(
                format!("The parser produce (lint) errors that weren't expected:\n {err}").into(),
            )
        }
    };

    let packages = source_info
        .packages_for_architecture(Architecture::X86_64)
        .collect::<Vec<MergedPackage>>();

    let json = serde_json::to_string_pretty(&packages)?;
    let name = case.file_stem().unwrap().to_str().unwrap().to_string();

    if json.contains("unexpected") {
        return Err(format!(
            "Found 'unexpected' keyword in json output. {}:\n{json}",
            "This indicates that data was included that shouldn't be in there"
        )
        .into());
    }

    if json.contains("beefc0ffee") {
        return Err(format!(
            "Found 'beefc0ffee' keyword in json output. {}:\n{json}",
            "This indicates that an checksum was included that shouldn't be in there"
        )
        .into());
    }

    // Compare the generated json with the expected snapshot.
    // Remove the usual module prefix by explicitly setting the snapshot path.
    // This is necessary, as we're manually sorting snapshots by test scenario.
    insta::with_settings!({
        snapshot_path => "correct_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, json);
    });

    Ok(())
}
