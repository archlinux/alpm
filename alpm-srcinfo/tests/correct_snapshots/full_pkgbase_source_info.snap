---
source: alpm-srcinfo/tests/correct.rs
description: full_pkgbase SourceInfo representation.
expression: source_info_json
---
{
  "base": {
    "name": "example",
    "description": "A example with all pkgbase properties set.",
    "url": "https://archlinux.org/",
    "changelog": "changelog.stub",
    "licenses": [
      "MIT",
      "Apache-2.0"
    ],
    "install": "install.sh.stub",
    "groups": [
      "group",
      "group_2"
    ],
    "options": [
      {
        "type": "package",
        "lto": false
      },
      {
        "type": "package",
        "emptydirs": true
      }
    ],
    "backups": [
      "etc/pacman.conf"
    ],
    "version": {
      "pkgver": "0.1.0",
      "pkgrel": {
        "major": 1,
        "minor": null
      },
      "epoch": 1
    },
    "pgp_fingerprints": [
      {
        "openpgp_v4_fingerprint": "4A0C4DFFC02E1A7ED969ED231C2358A25A10D94D"
      },
      {
        "openpgp_v4_fingerprint": "4A0C4DFFC02E1A7ED969ED231C2358A25A10D94E"
      }
    ],
    "architectures": [
      "x86_64",
      "aarch64"
    ],
    "architecture_properties": {
      "aarch64": {
        "dependencies": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "optional_dependencies": [
          {
            "package_relation": {
              "name": "unexpected",
              "version_requirement": null
            },
            "description": null
          }
        ],
        "provides": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "conflicts": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "replaces": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "check_dependencies": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "make_dependencies": [
          {
            "name": "unexpected",
            "version_requirement": null
          }
        ],
        "sources": [
          {
            "type": "File",
            "filename": null,
            "location": "unexpected"
          }
        ],
        "b2_checksums": [
          {
            "type": "Checksum",
            "digest": "beefc0ffee1882900bbcc29a84f8856a7bcd111979815435f9f7aa9c77fb77592fc56324ce3ba6c18df0609a5057e6e3589b9bcced9e7475f2984478b9028a97"
          }
        ],
        "md5_checksums": [],
        "sha1_checksums": [],
        "sha224_checksums": [
          {
            "type": "Checksum",
            "digest": "beefc0ffeec738367c4278c5028b5efcad1699b314535cbe2a0c34d5"
          }
        ],
        "sha256_checksums": [
          {
            "type": "Checksum",
            "digest": "beefc0ffee05bdc878082d735eb5306692a0db2410cd9463a8b4bb135b5b9562"
          }
        ],
        "sha384_checksums": [
          {
            "type": "Checksum",
            "digest": "beefc0ffee9363202624edbaf66538fd0749f735e7f28608cf8eebbf0ba196d76482021708b5831322b6846c44b73360"
          }
        ],
        "sha512_checksums": [
          {
            "type": "Checksum",
            "digest": "beefc0ffeebf31eb7c57a66bfb0f0ee5cef9398b5e4cc303468e08dae2702da55978402da94673e444f8c02754e94dedef4d12450319383c3a481d1c5cd90c82"
          }
        ]
      },
      "x86_64": {
        "dependencies": [
          {
            "name": "arch_default_dep",
            "version_requirement": null
          },
          {
            "name": "arch_default_dep_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "optional_dependencies": [
          {
            "package_relation": {
              "name": "arch_default_optdep",
              "version_requirement": null
            },
            "description": null
          },
          {
            "package_relation": {
              "name": "arch_default_optdep_2",
              "version_requirement": {
                "comparison": "Equal",
                "version": {
                  "pkgver": "2.0.0",
                  "epoch": null,
                  "pkgrel": null
                }
              }
            },
            "description": "With description"
          }
        ],
        "provides": [
          {
            "name": "arch_default_provides",
            "version_requirement": null
          },
          {
            "name": "arch_default_provides_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "conflicts": [
          {
            "name": "arch_default_conflict",
            "version_requirement": null
          },
          {
            "name": "arch_default_conflict_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "replaces": [
          {
            "name": "arch_default_replaces",
            "version_requirement": null
          },
          {
            "name": "arch_default_replaces_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "check_dependencies": [
          {
            "name": "arch_default_checkdep",
            "version_requirement": null
          },
          {
            "name": "arch_default_checkdep_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "make_dependencies": [
          {
            "name": "arch_default_makedep",
            "version_requirement": null
          },
          {
            "name": "arch_default_makedep_2",
            "version_requirement": {
              "comparison": "Equal",
              "version": {
                "pkgver": "2.0.0",
                "epoch": null,
                "pkgrel": null
              }
            }
          }
        ],
        "sources": [
          {
            "type": "File",
            "filename": null,
            "location": "arch_source_1.tar.gz"
          },
          {
            "type": "File",
            "filename": null,
            "location": "arch_source_2.patch"
          }
        ],
        "b2_checksums": [
          {
            "type": "Checksum",
            "digest": "87dd1c13611882900bbcc29a84f8856a7bcd111979815435f9f7aa9c77fb77592fc56324ce3ba6c18df0609a5057e6e3589b9bcced9e7475f2984478b9028a97"
          },
          {
            "type": "Skip"
          }
        ],
        "md5_checksums": [],
        "sha1_checksums": [],
        "sha224_checksums": [
          {
            "type": "Checksum",
            "digest": "8670e639dbc738367c4278c5028b5efcad1699b314535cbe2a0c34d5"
          },
          {
            "type": "Skip"
          }
        ],
        "sha256_checksums": [
          {
            "type": "Checksum",
            "digest": "53d397944805bdc878082d735eb5306692a0db2410cd9463a8b4bb135b5b9562"
          },
          {
            "type": "Skip"
          }
        ],
        "sha384_checksums": [
          {
            "type": "Checksum",
            "digest": "5487a70ec49363202624edbaf66538fd0749f735e7f28608cf8eebbf0ba196d76482021708b5831322b6846c44b73360"
          },
          {
            "type": "Skip"
          }
        ],
        "sha512_checksums": [
          {
            "type": "Checksum",
            "digest": "1816c57b4abf31eb7c57a66bfb0f0ee5cef9398b5e4cc303468e08dae2702da55978402da94673e444f8c02754e94dedef4d12450319383c3a481d1c5cd90c82"
          },
          {
            "type": "Skip"
          }
        ]
      }
    },
    "dependencies": [
      {
        "name": "default_dep",
        "version_requirement": null
      },
      {
        "name": "default_dep_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "optional_dependencies": [
      {
        "package_relation": {
          "name": "default_optdep",
          "version_requirement": null
        },
        "description": null
      },
      {
        "package_relation": {
          "name": "default_optdep_2",
          "version_requirement": {
            "comparison": "Equal",
            "version": {
              "pkgver": "2.0.0",
              "epoch": null,
              "pkgrel": null
            }
          }
        },
        "description": "With description"
      }
    ],
    "provides": [
      {
        "name": "default_provides",
        "version_requirement": null
      },
      {
        "name": "default_provides_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "conflicts": [
      {
        "name": "default_conflict",
        "version_requirement": null
      },
      {
        "name": "default_conflict_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "replaces": [
      {
        "name": "default_replaces",
        "version_requirement": null
      },
      {
        "name": "default_replaces_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "check_dependencies": [
      {
        "name": "default_checkdep",
        "version_requirement": null
      },
      {
        "name": "default_checkdep_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "make_dependencies": [
      {
        "name": "default_makedep",
        "version_requirement": null
      },
      {
        "name": "default_makedep_2",
        "version_requirement": {
          "comparison": "Equal",
          "version": {
            "pkgver": "2.0.0",
            "epoch": null,
            "pkgrel": null
          }
        }
      }
    ],
    "sources": [
      {
        "type": "File",
        "filename": null,
        "location": "source_1.tar.gz"
      },
      {
        "type": "SourceUrl",
        "filename": "source_2",
        "source_url": {
          "url": "https://upstream.com/example.git",
          "vcs_info": {
            "protocol": "git",
            "fragment": {
              "branch": "0.1.0"
            },
            "signed": true
          }
        }
      },
      {
        "type": "File",
        "filename": null,
        "location": "source_3.zip"
      },
      {
        "type": "File",
        "filename": null,
        "location": "source_4.tar.gz"
      }
    ],
    "no_extracts": [
      "source_3.zip",
      "source_4.tar.gz"
    ],
    "b2_checksums": [
      {
        "type": "Checksum",
        "digest": "87dd1c13611882900bbcc29a84f8856a7bcd111979815435f9f7aa9c77fb77592fc56324ce3ba6c18df0609a5057e6e3589b9bcced9e7475f2984478b9028a97"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      }
    ],
    "md5_checksums": [],
    "sha1_checksums": [],
    "sha224_checksums": [
      {
        "type": "Checksum",
        "digest": "8670e639dbc738367c4278c5028b5efcad1699b314535cbe2a0c34d5"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      }
    ],
    "sha256_checksums": [
      {
        "type": "Checksum",
        "digest": "53d397944805bdc878082d735eb5306692a0db2410cd9463a8b4bb135b5b9562"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      }
    ],
    "sha384_checksums": [
      {
        "type": "Checksum",
        "digest": "5487a70ec49363202624edbaf66538fd0749f735e7f28608cf8eebbf0ba196d76482021708b5831322b6846c44b73360"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      }
    ],
    "sha512_checksums": [
      {
        "type": "Checksum",
        "digest": "1816c57b4abf31eb7c57a66bfb0f0ee5cef9398b5e4cc303468e08dae2702da55978402da94673e444f8c02754e94dedef4d12450319383c3a481d1c5cd90c82"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      },
      {
        "type": "Skip"
      }
    ]
  },
  "packages": [
    {
      "name": "example",
      "description": {
        "override": "No"
      },
      "url": {
        "override": "No"
      },
      "changelog": {
        "override": "No"
      },
      "licenses": {
        "override": "No"
      },
      "install": {
        "override": "No"
      },
      "groups": {
        "override": "No"
      },
      "options": {
        "override": "No"
      },
      "backups": {
        "override": "No"
      },
      "architectures": null,
      "architecture_properties": {},
      "dependencies": {
        "override": "No"
      },
      "optional_dependencies": {
        "override": "No"
      },
      "provides": {
        "override": "No"
      },
      "conflicts": {
        "override": "No"
      },
      "replaces": {
        "override": "No"
      }
    }
  ]
}
